partiql-lang/src/main/kotlin/org/partiql/lang/syntax/impl/PartiQLPigVisitor.kt

```diff
@@ -1479,11 +1520,13 @@ internal class PartiQLPigVisitor(val customTypes: List<CustomType> = listOf(), p
         else -> ctx.map { visitOrNull(it, clazz) }
     }
 
+    // TODO: Inspect `visitOrNull` call sites and replace with direct `visitXxx` calls or 'as' casts, as appropriate.
     private fun <T : PartiqlAst.PartiqlAstNode> visitOrNull(ctx: ParserRuleContext?, clazz: KClass<T>): T? = when (ctx) {
         null -> null
         else -> clazz.cast(visit(ctx))
     }
```

Comment:
Nice PR. Thank you for the cleaning effort! This has some background. The visitOrNull with the reflection based cast came from a similar Java implementation.

We can improve the readability (and use the `as` cast) with a type reified inline. [Here's what I did for the new implementation.
](https://github.com/partiql/partiql-lang-kotlin/blob/d353514d147f60b2c4ff207c55f69b65acfd73c0/partiql-parser/src/main/kotlin/org/partiql/parser/impl/PartiQLParserDefault.kt#L1839)

These inline functions are like macros

```suggestion
private inline fun <reified T : AstNode> visitOrNull(ctx: ParserRuleContext?): T? = ctx?.let { visit(it) as T }
```


Then we can replace the current `visitOrNull` to achieve the same affect. Additionally, it would be nice to pass a method reference for which visit to call. But at that point, we should just do the explicit visit method.

Reply:
Thanks for pointing out this inline+reified language construct!  With it, my major concern (about reflection) goes away.  Then, going with reified visit/visitOrNull or not comes down to a matter of preference, I think.  

In my opinion, the bodies of these functions are so trivial that I do not feel the abstraction is worth it.  The call sites do not give much mental savings either: 
```
   visit(ctx.exprBagOp()) as PartiqlAst.Expr
```  
vs
```
   visit<PartiqlAst.Expr>(ctx.exprBagOp())
```

Regardless the above, do you I read it right you support doing 
```
   ctx.whereClause()?.let { visitWhereClause(it) }
```
when a specific visitor is available?

Reply:
Yes, `ctx.whereClause()?.let { visitWhereClause(it) }` when we know the type would be preferable. I recall only a few places where we didn't now the type .. meaning the visitOrNull is becoming less useful as time goes on